src/backend/distributed/master/master_repair_shards.c

```diff
@@ -37,133 +39,239 @@
-	ShardPlacement *targetPlacement = NULL;
-	WorkerNode *targetNode = NULL;
-	List *ddlCommandList = NIL;
-	bool dataCopied = false;
-	char relationKind = '\0';
+	foreach(colocatedTableCell, colocatedTableList)
+	{
+		Oid colocatedTableId = lfirst_oid(colocatedTableCell);
+		char relationKind = '\0';
 
-	EnsureTableOwner(distributedTableId);
+		/* check that user has owner rights in all co-located tables */
+		EnsureTableOwner(colocatedTableId);
 
-	/*
-	 * By taking an exclusive lock on the shard, we both stop all modifications
-	 * (INSERT, UPDATE, or DELETE) and prevent concurrent repair operations from
-	 * being able to operate on this shard.
-	 */
-	LockShardResource(shardId, ExclusiveLock);
+		relationKind = get_rel_relkind(colocatedTableId);
+		if (relationKind == RELKIND_FOREIGN_TABLE)
+		{
+			char *relationName = get_rel_name(colocatedTableId);
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot repair shard"),
+							errdetail("Table %s is a foreign table. Repairing "
+									  "shards backed by foreign tables is "
+									  "not supported.", relationName)));
+		}
+	}
+
+	/* we sort colocatedShardList so that lock operations will not cause any deadlocks */
+	colocatedShardList = SortList(colocatedShardList, CompareShardIntervalsById);
 
 	/*
-	 * We've stopped data modifications of this shard, but we plan to move
-	 * a placement to the healthy state, so we need to grab a shard metadata
-	 * lock (in exclusive mode) as well.
+	 * FIXME
+	 * To reviewer, I am not sure about locking structure here;
+	 * Are these correct lock functions?
+	 * Should we take lock on relationId instead of shardId?
+	 * Should we take lock on colocationId?
 	 */
-	LockShardDistributionMetadata(shardId, ExclusiveLock);
-
-	relationOwner = TableOwner(distributedTableId);
-
-	shardPlacementList = ShardPlacementList(shardId);
-	sourcePlacement = SearchShardPlacementInList(shardPlacementList, sourceNodeName,
-												 sourceNodePort);
-	if (sourcePlacement->shardState != FILE_FINALIZED)
+	foreach(colocatedShardCell, colocatedShardList)
 	{
-		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-						errmsg("source placement must be in finalized state")));
+		ShardInterval *colocatedShard = (ShardInterval *) lfirst(colocatedShardCell);
+		uint64 colocatedShardId = colocatedShard->shardId;
+
+		/*
+		 * We've stopped data modifications of this shard, but we plan to move
+		 * a placement to the healthy state, so we need to grab a shard metadata
+		 * lock (in exclusive mode) as well.
+		 */
+		LockShardDistributionMetadata(colocatedShardId, ExclusiveLock);
+
+		/*
+		 * doRepair means that we should drop placement from target before recreating it.
+		 * It should be set to true if we want to repair existing shard placement. If it
+		 * is false, we create new shard placement without trying to delete anything.
+		 */
+		if (doRepair)
+		{
+			EnsureShardCanBeRepaired(colocatedShardId, sourceNodeName, sourceNodePort,
```

Comment:
Thinking about this, this is not a realistic scenario until we have the logic to mark all colocated placements invalid (i.e. #810 )
In a realistic case, only a single shard placement should be on a situation where it needs to be repaired. Since `EnsureShardCanBeRepaired ` will fail its co-located placements, the user wouldn't be able to repair the shard at all.

Since we expect #810 to be considered just after 6.0, lets keep the co-located repair logic here, but, with a more relaxed checks.

@marcocitus what do you think about this?